File: apps/dashboard/src/lib/asr-socket.ts
'use client';

// Realtime voice input, browser half: the socket to /api/asr/<sessionId> and the
// little state machine that turns its events into text you can put on screen.
//
// THREE LAYERS OF TEXT, and keeping them apart is the whole design (see
// docs/realtime-voice-input-design.md):
//
//   partial  the sentence being spoken RIGHT NOW. Rewrites itself wholesale every
//            few hundred ms — "发" → "发 red hot" → "把Red Hole的隧道重启" — which
//            is exactly the effect worth showing, so it is passed through
//            unsmoothed.
//
//   tail     sentences the server has closed, joined.
//
//   polished each of those sentences again, corrected, replacing itself in place.
//            Arrives out of order (the jobs run concurrently), so segments are
//            addressed by id and the tail is rebuilt from the array — never
//            patched by string offset.
//
// All three end up in the composer's textarea: the words appear where you are
// going to send them, and the correction happens in place. They are reported
// separately rather than pre-joined for one reason — when the socket dies, only
// the CLOSED sentences may stay in the draft. The partial's audio is about to be
// re-transcribed by the batch fallback, and leaving it would deliver it twice.
//
// The socket carries the dashboard key as a `hermit-key.<token>` subprotocol,
// same as the terminal, and 401s for a scoped agent-share token — the caller
// treats that like any other failure and stays on press-and-hold.

import { getActiveKey } from '@/lib/keyring';
import { wsUrl } from '@/lib/api-base';
import {
  asrInitial, asrState, asrStep, type AsrModel, type DictationState,
} from '@/lib/asr-reduce';

export type { DictationState };

export interface AsrSocketEvents {
  /** The ASR task is live; audio is being transcribed. */
  onReady: () => void;
  /** Any of the three layers changed. */
  onState: (state: DictationState) => void;
  /** A sentence closed — safe to drop the capture layer's fallback buffer. */
  onSentence: () => void;
  /** stop() finished and everything landed. `tail` is final. */
  onDone: (tail: string) => void;
  /** The socket is unusable. Caller falls back to the batch route. */
  onFailure: (message: string) => void;
}

export interface AsrSocket {
  /** Queue 16 kHz mono PCM16 (buffered until the socket opens). */
  send: (pcm: Int16Array) => void;
  /** Ask the server to close the ASR task and flush; resolves via onDone. */
  stop: () => void;
  /** Drop it now, no flush, no events. */
  close: () => void;
}

/** Audio held while the socket is still opening — ~4 s, well past a normal connect. */
const PREOPEN_MAX_BYTES = 4 * 16_000 * 2;
/** stop() waits this long for the tail + corrections before giving up on them. */
const DONE_TIMEOUT_MS = 12_000;

export function openAsrSocket(sessionId: string, events: AsrSocketEvents): AsrSocket {
  // Follows the active keyring entry's deployment; see lib/api-base.ts.
  const url = wsUrl(`/api/asr/${encodeURIComponent(sessionId)}`);

  // Everything the frames say, and nothing about the socket — see
  // `lib/asr-reduce.ts`, which iOS runs the same table over.
  let model: AsrModel = asrInitial();
  let dead = false;
  let stopping = false;
  let doneTimer: ReturnType<typeof setTimeout> | null = null;
  let preopen: Int16Array[] = [];
  let preopenBytes = 0;

  const tail = () => asrState(model).tail;
  const emit = () => {
    if (dead) return;
    events.onState(asrState(model));
  };

  const sock = new WebSocket(url, [`hermit-key.${getActiveKey()}`]);
  sock.binaryType = 'arraybuffer';

  const fail = (message: string) => {
    if (dead) return;
    dead = true;
    if (doneTimer) { clearTimeout(doneTimer); doneTimer = null; }
    events.onFailure(message);
    try { sock.close(); } catch { /* already gone */ }
  };

  const settleDone = () => {
    if (dead) return;
    dead = true;
    if (doneTimer) { clearTimeout(doneTimer); doneTimer = null; }
    events.onDone(tail());
    try { sock.close(); } catch { /* already gone */ }
  };

  sock.onopen = () => {
    for (const c of preopen) {
      try { sock.send(c.buffer as ArrayBuffer); } catch { /* closing */ }
    }
    preopen = [];
    preopenBytes = 0;
  };

  sock.onmessage = (ev) => {
    const step = asrStep(model, String(ev.data));
    // A frame the reducer ignored — unknown type, unparseable, a correction for
    // a sentence that was never opened — hands the SAME object back, and
    // redrawing on it would write the identical tail into the draft again.
    const changed = step.model !== model;
    model = step.model;
    switch (step.effect.kind) {
      case 'ready':    events.onReady(); return;
      case 'sentence': emit(); events.onSentence(); return;
      case 'done':     emit(); settleDone(); return;
      case 'fail':     fail(step.effect.message); return;
      default:         if (changed) emit(); return;
    }
  };

  // A close we didn't ask for is a failure; one during stop() means the server
  // hung up before 'done' — the sentences already closed are still good, so that
  // is a clean finish, not a loss.
  sock.onclose = () => {
    if (dead) return;
    if (stopping) settleDone();
    else fail('connection closed');
  };
  sock.onerror = () => { if (!dead && !stopping) fail('connection error'); };

  return {
    send(pcm: Int16Array) {
      if (dead || stopping) return;
      if (sock.readyState === WebSocket.CONNECTING) {
        // The mic is already live while the socket handshakes; hold the first
        // syllables rather than clipping them.
        preopen.push(pcm);
        preopenBytes += pcm.byteLength;
        while (preopenBytes > PREOPEN_MAX_BYTES && preopen.length > 1) {
          preopenBytes -= preopen.shift()!.byteLength;
        }
        return;
      }
      if (sock.readyState !== WebSocket.OPEN) return;
      try { sock.send(pcm.buffer as ArrayBuffer); } catch { /* closing */ }
    },

    stop() {
      if (dead || stopping) return;
      stopping = true;
      if (sock.readyState !== WebSocket.OPEN) { settleDone(); return; }
      try { sock.send(JSON.stringify({ type: 'stop' })); } catch { settleDone(); return; }
      doneTimer = setTimeout(settleDone, DONE_TIMEOUT_MS);
    },

    close() {
      dead = true;
      if (doneTimer) { clearTimeout(doneTimer); doneTimer = null; }
      try { sock.close(); } catch { /* already gone */ }
    },
  };
}

File: apps/ios/Hermit/AsrSocket.swift
import Foundation

/// The socket to `/api/asr/<sessionId>`, and nothing else.
///
/// The native half of `apps/dashboard/src/lib/asr-socket.ts` — minus everything
/// that decides what the words ARE, which is `DictationCore.swift` and is run
/// against the web's own answers by `tools/hold-fixture.sh`. What is left here
/// is the transport: the handshake, the pre-open buffer, and the two timeouts.
///
/// ## Auth is a SUBPROTOCOL, not a header
///
/// The shell could set `x-asst-key` on this request — it is not a browser and
/// nothing stops it. The server does not read it: `/api/asr` authenticates the
/// `hermit-key.<token>` entry in `Sec-WebSocket-Protocol`, because the browser
/// half cannot set a header on a WebSocket and that is the only channel it has.
/// Sending the header instead would 401 with a message about a missing key while
/// the key sat in the request.
///
/// ## Every callback lands on the main queue
///
/// `URLSessionWebSocketTask` delivers on its own queue, and everything upstream
/// of this writes into a draft the keyboard is typing into. Hopping here means
/// no caller has to remember to.
final class AsrSocket {
    struct Events {
        /// The ASR task is live; audio is being transcribed.
        var onReady: () -> Void = {}
        /// Any of the three layers changed.
        var onState: (DictationState) -> Void = { _ in }
        /// A sentence closed — the capture layer's fallback buffer can go.
        var onSentence: () -> Void = {}
        /// `stop()` finished and everything landed. The tail is final.
        var onDone: (String) -> Void = { _ in }
        /// The socket is unusable. The caller falls back, or gives up.
        var onFailure: (String) -> Void = { _ in }
    }

    /// Audio held while the socket is still opening — ~4 s, well past a normal
    /// connect. The mic is already live during the handshake; holding the first
    /// syllables beats clipping them.
    private static let preopenMaxBytes = 4 * 16_000 * 2
    /// `stop()` waits this long for the tail and its corrections.
    private static let doneTimeout: TimeInterval = 12

    private let task: URLSessionWebSocketTask
    private let session: URLSession
    private let events: Events

    private var model = AsrModel()
    private var dead = false
    private var stopping = false
    private var open = false
    private var preopen: [Data] = []
    private var preopenBytes = 0
    private var doneTimer: Timer?

    /// - Parameters:
    ///   - root: scheme + host + port, as `HermitAPI.rootOf` produces.
    ///   - key: the machine key, read once — a socket cannot re-authenticate
    ///     mid-stream, so unlike `HermitAPI` there is nothing to re-read.
    init?(root: String, sessionId: String, key: String, events: Events,
          session: URLSession = .shared) {
        guard var c = URLComponents(string: root) else { return nil }
        c.scheme = c.scheme == "http" ? "ws" : "wss"
        c.path = "/api/asr/" + sessionId
        guard let url = c.url else { return nil }

        self.session = session
        self.events = events
        self.task = session.webSocketTask(with: url, protocols: ["hermit-key.\(key)"])
        task.resume()
        receive()
    }

    // MARK: - Sending

    /// Queue 16 kHz mono PCM16. Buffered until the socket opens.
    func send(_ pcm: Data) {
        guard !dead, !stopping else { return }
        guard open else {
            preopen.append(pcm)
            preopenBytes += pcm.count
            while preopenBytes > Self.preopenMaxBytes && preopen.count > 1 {
                preopenBytes -= preopen.removeFirst().count
            }
            return
        }
        task.send(.data(pcm)) { _ in /* a write that fails is a close, and the read sees it */ }
    }

    /// Ask the server to close the ASR task and flush. Resolves via `onDone`.
    func stop() {
        guard !dead, !stopping else { return }
        stopping = true
        let frame = #"{"type":"stop"}"#
        task.send(.string(frame)) { [weak self] error in
            guard let self else { return }
            DispatchQueue.main.async {
                if error != nil { self.settleDone(); return }
                self.doneTimer = Timer.scheduledTimer(withTimeInterval: Self.doneTimeout,
                                                      repeats: false) { [weak self] _ in
                    self?.settleDone()
                }
            }
        }
    }

    /// Drop it now: no flush, no events.
    func close() {
        dead = true
        doneTimer?.invalidate()
        doneTimer = nil
        task.cancel(with: .goingAway, reason: nil)
    }

    // MARK: - Receiving

    private func receive() {
        task.receive { [weak self] result in
            guard let self else { return }
            DispatchQueue.main.async {
                switch result {
                case let .success(message):
                    // The first message is proof the handshake finished, which is
                    // the only "open" signal `URLSessionWebSocketTask` offers
                    // without a delegate. The server sends `ready` unprompted, so
                    // this is never a long wait.
                    self.flushPreopen()
                    if case let .string(text) = message { self.apply(text) }
                    if case let .data(bytes) = message,
                       let text = String(data: bytes, encoding: .utf8) { self.apply(text) }
                    if !self.dead { self.receive() }
                case .failure:
                    // A close we did not ask for is a failure; one during stop()
                    // means the server hung up before `done` — the sentences
                    // already closed are still good, so that is a clean finish.
                    if self.dead { return }
                    if self.stopping { self.settleDone() } else { self.fail("connection closed") }
                }
            }
        }
    }

    private func flushPreopen() {
        guard !open else { return }
        open = true
        for chunk in preopen { task.send(.data(chunk)) { _ in } }
        preopen = []
        preopenBytes = 0
    }

    private func apply(_ raw: String) {
        guard !dead else { return }
        let step = AsrReduce.step(model, raw)
        // A frame the reducer ignored hands the same value back, and redrawing on
        // it would write the identical tail into the draft again. Same test the
        // web makes, for the same reason — see `lib/asr-reduce.ts`.
        let changed = step.model != model
        model = step.model
        switch step.effect {
        case .ready:
            events.onReady()
        case .sentence:
            emit()
            events.onSentence()
        case .done:
            emit()
            settleDone()
        case let .fail(message):
            fail(message)
        case .none:
            if changed { emit() }
        }
    }

    private func emit() {
        guard !dead else { return }
        events.onState(AsrReduce.state(model))
    }

    private func settleDone() {
        guard !dead else { return }
        dead = true
        doneTimer?.invalidate()
        doneTimer = nil
        events.onDone(AsrReduce.state(model).tail)
        task.cancel(with: .normalClosure, reason: nil)
    }

    private func fail(_ message: String) {
        guard !dead else { return }
        dead = true
        doneTimer?.invalidate()
        doneTimer = nil
        events.onFailure(message)
        task.cancel(with: .goingAway, reason: nil)
    }
}
